src/nlsolve/newton.jl

```diff
@@ -119,31 +112,37 @@ end
   f = nlsolve_f(integrator)
   isdae = f isa DAEFunction
 
-  if !isdae
-    mass_matrix = integrator.f.mass_matrix
+  if DiffEqBase.has_destats(integrator)
+    integrator.destats.nf += 1
   end
 
   if isdae
     @.. ztmp = (tmp + α * z) * invγdt
     @.. ustep = uprev + z
     f(k, ztmp, ustep, p, tstep)
-  else
-    @.. ustep = tmp + γ * z
-    f(k, ustep, p, tstep)
-  end
-  if DiffEqBase.has_destats(integrator)
-    integrator.destats.nf += 1
-  end
-
-  if isdae
     b = vec(k)
   else
-    if mass_matrix === I
-      @.. ztmp = (dt * k - z) * invγdt
+    mass_matrix = integrator.f.mass_matrix
+    if isodemultistep(unwrap_alg(integrator, true))
+      ustep = z
+      f(k, z, p, tstep)
+      if mass_matrix === I
+        @.. ztmp = tmp .+ k - α * z .* invγdt
```

Comment:
I guess since we use `@..` the additional broadcasts are not needed?
```suggestion
        @.. ztmp = tmp + k - α * z * invγdt
```

Reply:
Maybe one could precompute `α * invγdt` as in the out-of-place version.